enhancements/agent-installer/simplified-registry-less-cluster-operations.md

```diff
@@ -0,0 +1,426 @@
+   downloaded ISO.
+   (note: even though not strictly relevant for the current proposal,
+   we'll assume that the deployment configuration details are already
+   available in the installation environment - either provided statically or
+   interactively by the user - such as the rendezvous node selection and cluster
+   specifications).
+4. The installation process starts automatically and runs unattended until
+   completed.
+
+#### Upgrade
+
+1. The user downloads the extended RHCOS ISO from the Red Hat Customer Portal,
+   by selecting a compatible newer version release for the target cluster.
+2. The user moves the extended RHCOS ISO into the disconnected environment.
+3. The user attaches the extended RHCOS ISO to one of the control plane nodes
+   of the target cluster.
+4. The user edits the IRI resource by adding the release image pullspec to the
+   `spec.releases` field.
+5. The user waits until the copy process of the new release image is completed
+   for all the control plane nodes. The progress and status of the task could
+   be monitored via the IRI `status` field.
+6. Once completed, the user upgrades the cluster via the `oc adm upgrade
+   --allow-explicit-upgrade --to-image=<release-image>`
+
+#### Post-upgrade optional steps
+
+1. The user edits the IRI resource by removing the release image pullspec not
+   in use anymore.
+2. The specified release image is removed from all the control plane nodes. The
+   progress and status of the task could be monitored via the IRI `status`
+   field.
+
+### Node expansion
+
+1. The user runs the `oc adm node-image create` command from within the
+   disconnected environment (with optionally some additional node
+   configuration).
+2. The user attaches the newly generate node ISO to the target machine(s),
+   and boots it. The progress and status of the task can be monitored via the
+   `oc adm node-image monitor` command.
+3. The user waits until the CSRs are generated, and manually approves them
+   via the `oc adm certificate approve` command.
+   
+### API Extensions
+
+#### InternalReleaseImage custom resource
+
+The `InternalReleaseImage` is a new singleton custom resource, controlled by
+MCO, used to keep track and manage the OCP release images stored into the
+control planes nodes. In particular, it will be used for:
+
+* Triggering a new release image copy task - as a preliminary upgrade step
+* Delete from disk a release image version not anymore in use - as an optional
+  post-upgrade step
+* Enabled/disable the registry services
+
+This is an example of how the CR will look like:
+
+```
+apiVersion: releases.openshift.io/v1alpha1
+kind: InternalReleaseImage
+metadata:
+  name: cluster
+spec:
+  releases:
+    - "quay.io/openshift-release-dev/ocp-release:4.18.0-x86_64"
+    - "quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64"
+  enabled: true
+status:
+  releases:
+    - "quay.io/openshift-release-dev/ocp-release:4.18.0-x86_64"
```

Comment:
Thinking about what we should report here, I wonder if each individual status entry for each payload should have subfields around progressing/success/error/etc, instead of overall conditions, so it's easier to see which payloads are there, which are being pulled, etc.

Although since a user is unlikely to specify more than 2 payloads here at a time, perhaps we just report the most recent error? Would the MCD would be constantly reconciling all releases it expects to be on disk? Or just the latest added/removed one?

Reply:
In that case I think the status should have sub sections per release on the status. i.e. it would be nice to see how many nodes have successfully pulled each release, what errors there are, etc.